File: library/src/main/java/com/afollestad/aesthetic/views/AestheticTextView.java
package com.afollestad.aesthetic.views;

import android.content.Context;
import android.support.annotation.RestrictTo;
import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;

import com.afollestad.aesthetic.Aesthetic;

import rx.Observable;
import rx.Subscription;

import static android.support.annotation.RestrictTo.Scope.LIBRARY_GROUP;
import static com.afollestad.aesthetic.Rx.distinctToMainThread;
import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;

@RestrictTo(LIBRARY_GROUP)
public class AestheticTextView extends AppCompatTextView {

  private Subscription subscription;

  public AestheticTextView(Context context) {
    super(context);
  }

  public AestheticTextView(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public AestheticTextView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    String idName;
    try {
      idName = getResources().getResourceName(getId());
    } catch (Exception e) {
      idName = null;
    }

    // TODO when to use inverse colors?

    Observable<Integer> obs;
    if (idName == null || idName.contains("title") || idName.contains("header")) {
      obs = Aesthetic.get().primaryTextColor();
    } else {
      obs = Aesthetic.get().secondaryTextColor();
    }
    subscription =
        obs.compose(distinctToMainThread()).subscribe(this::setTextColor, onErrorLogAndRethrow());
  }

  @Override
  protected void onDetachedFromWindow() {
    subscription.unsubscribe();
    super.onDetachedFromWindow();
  }
}

File: library/src/main/java/com/afollestad/aesthetic/views/AestheticTextView.kt
/*
 * Licensed under Apache-2.0
 *
 * Designed and developed by Aidan Follestad (@afollestad)
 */
package com.afollestad.aesthetic.views

import android.content.Context
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import com.afollestad.aesthetic.Aesthetic.Companion.get
import com.afollestad.aesthetic.internal.AttrWizard
import com.afollestad.aesthetic.utils.distinctToMainThread
import com.afollestad.aesthetic.utils.observableForAttrName
import com.afollestad.aesthetic.utils.subscribeTextColor
import com.afollestad.aesthetic.utils.unsubscribeOnDetach

/** @author Aidan Follestad (afollestad) */
class AestheticTextView(
  context: Context,
  attrs: AttributeSet? = null
) : AppCompatTextView(context, attrs) {

  private val wizard = AttrWizard(context, attrs)
  private val textColorValue = wizard.getRawValue(android.R.attr.textColor)

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()

    val defaultObs = if (id == android.R.id.title) {
      get().textColorPrimary()
    } else {
      get().textColorSecondary()
    }
    get().observableForAttrName(textColorValue, defaultObs)
        ?.distinctToMainThread()
        ?.subscribeTextColor(this)
        ?.unsubscribeOnDetach(this)
  }
}
